internal/handlers/pg/pgdb/query.go

```diff
@@ -246,20 +247,42 @@ func prepareWhereClause(sqlFilters *types.Document) (string, []any) {
 	var p Placeholder
 
 	for k, v := range sqlFilters.Map() {
-		switch k {
-		case "_id":
-			switch v := v.(type) {
-			case string:
-				filters = append(filters, fmt.Sprintf(`((_jsonb->'_id')::jsonb = %s)`, p.Next()))
-				args = append(args, string(must.NotFail(pjson.MarshalSingleValue(v))))
+		if len(k) != 0 && (k[0] == '$' || types.NewPathFromString(k).Len() > 1) {
+			// TODO $eq and $ne https://github.com/FerretDB/FerretDB/issues/1840
+			// TODO dot notation https://github.com/FerretDB/FerretDB/issues/1841
+			continue
+		}
 
-			case types.ObjectID:
+		// don't iterate through array for _id keys to simplify the query
+		if k == "_id" {
+			switch v := v.(type) {
+			case string, types.ObjectID:
 				filters = append(filters, fmt.Sprintf(`((_jsonb->'_id')::jsonb = %s)`, p.Next()))
 				args = append(args, string(must.NotFail(pjson.MarshalSingleValue(v))))
 			}
-		default:
+
 			continue
 		}
+
+		switch v := v.(type) {
+		case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+			continue
+
+		case float64, string, types.ObjectID, int32, int64:
+			sql := fmt.Sprintf(
+				`((_jsonb->%[1]s)::jsonb = %[2]s)`+ // Select if value under the key k is equal to value v.
+					// If it's not, but the value under the key k is an array - select if it contains the value equal to v.
+					` OR (jsonb_typeof(_jsonb->%[1]s) = 'array' AND`+
+					` EXISTS (SELECT 1 FROM jsonb_array_elements(_jsonb->%[1]s) as elem WHERE (elem::jsonb = %[2]s)))`,
```

Comment:
It think that could be simplified using `@>` operator. See https://www.postgresql.org/docs/current/datatype-json.html#JSON-CONTAINMENT
> SELECT '["foo", "bar"]'::jsonb @> '"bar"'::jsonb;

Reply:
Looks much simpler!